powerauth-push-server/src/main/java/io/getlime/push/service/DeviceRegistrationService.java

```diff
@@ -0,0 +1,262 @@
+                    // and create a new row.
+                    devices.stream().filter(existingDevice -> !usedDeviceRegistrationIds.contains(existingDevice.getId())).forEach(pushDeviceRepository::delete);
+                    device = initDeviceRegistrationEntity(appCredentials, pushToken);
+                }
+                device.setTimestampLastRegistered(new Date());
+                device.setPlatform(convert(platform));
+                updateActivationForDevice(device, activationId);
+                PushDeviceRegistrationEntity registeredDevice = pushDeviceRepository.save(device);
+                usedDeviceRegistrationIds.add(registeredDevice.getId());
+            } catch (PushServerException ex) {
+                logger.error(ex.getMessage(), ex);
+                registrationFailed.set(true);
+            }
+        });
+
+        if (registrationFailed.get()) {
+            throw new PushServerException("Device registration failed");
+        }
+    }
+
+    public void updateStatus(final UpdateDeviceStatusRequest request) throws PushServerException {
+        final String activationId = request.getActivationId();
+
+        final List<PushDeviceRegistrationEntity> device = pushDeviceRepository.findByActivationId(activationId);
+
+        final ActivationStatus activationStatus = request.getActivationStatus() == null ? fetchActivationStatus(activationId) : request.getActivationStatus();
+
+        for (PushDeviceRegistrationEntity registration : device) {
+            registration.setActive(activationStatus == ActivationStatus.ACTIVE);
+            pushDeviceRepository.save(registration);
+        }
+    }
+
+    public void delete(final String appId, final String pushToken) {
+        final List<PushDeviceRegistrationEntity> devices = pushDeviceRepository.findByAppCredentialsAppIdAndPushToken(appId, pushToken);
+        pushDeviceRepository.deleteAll(devices);
+    }
+
+    /**
+     * Lookup device registrations using app ID, activation ID and push token.
+     * <br/>
+     * The query priorities are ranging from most exact to least exact match:
+     * <ul>
+     *     <li>Lookup by activation ID and push token.</li>
+     *     <li>Lookup by activation ID.</li>
+     *     <li>Lookup by application ID and push token.</li>
+     * </ul>
+     * @param appId Application ID.
+     * @param activationId Activation ID.
+     * @param pushToken Push token.
+     * @return List of found device registration entities.
+     */
+    private List<PushDeviceRegistrationEntity> lookupDeviceRegistrations(String appId, String activationId, String pushToken) throws PushServerException {
+        List<PushDeviceRegistrationEntity> deviceRegistrations;
+        // At first, lookup the device registrations by match on activationId and pushToken.
+        deviceRegistrations = pushDeviceRepository.findByActivationIdAndPushToken(activationId, pushToken);
```

Comment:
It could be inlined with the declaration. Moreover rewriting to `final` would be less error-prone.